instrumentation/action_mailer/lib/opentelemetry/instrumentation/action_mailer/instrumentation.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActionMailer
+      # The Instrumentation class contains logic to detect and install the ActionMailer instrumentation
+      class Instrumentation < OpenTelemetry::Instrumentation::Base
+        MINIMUM_VERSION = Gem::Version.new('6.1.0')
+        install do |_config|
+          resolve_email_address
+          ecs_mail_convention
+          require_dependencies
+        end
+
+        present do
+          defined?(::ActionMailer)
+        end
+
+        compatible do
+          gem_version >= MINIMUM_VERSION
+        end
+
+        option :disallowed_notification_payload_keys, default: [], validate: :array
+        option :notification_payload_transform,       default: nil, validate: :callable
+        option :email_address,                        default: :omit, validate: %I[omit include]
+
+        private
+
+        def gem_version
+          ::ActionMailer.version
+        end
+
+        def resolve_email_address
+          return unless ActionMailer::Instrumentation.instance.config[:email_address] == :omit
+
+          ActionMailer::Instrumentation.instance.config[:disallowed_notification_payload_keys] += ['email.to.address', 'email.from.address', 'email.cc.address', 'email.bcc.address']
+        end
+
+        def ecs_mail_convention
+          if ActionMailer::Instrumentation.instance.config[:notification_payload_transform].nil?
+            transform_attributes = ->(payload) { transform_payload(payload) }
+          else
+            original_callable = ActionMailer::Instrumentation.instance.config[:notification_payload_transform]
+            transform_attributes = lambda do |payload|
+              new_payload = transform_payload(payload)
+              user_payload = original_callable.call(new_payload)
+              if user_payload.instance_of?(Hash)
+                user_payload
+              else
+                OpenTelemetry.logger.error("ActionMailer: transformed payload is #{user_payload.class} (require Hash)")
+                new_payload
+              end
+            end
+          end
+          ActionMailer::Instrumentation.instance.config[:notification_payload_transform] = transform_attributes
+        end
+
+        # email attribute key convention is obtained from: https://www.elastic.co/guide/en/ecs/8.11/ecs-email.html
+        def transform_payload(payload)
+          new_payload = {
+            'email.message_id' => payload[:message_id],
+            'email.subject' => payload[:subject],
+            'email.x_mailer' => payload[:mailer],
+            'email.to.address' => payload[:to],
+            'email.from.address' => payload[:from],
+            'email.cc.address' => payload[:cc],
+            'email.bcc.address' => payload[:bcc],
+            'email.origination_timestamp' => payload[:date]
+          }
+          new_payload.compact!
```

Comment:
This is a micro optimization but checking if the values are `nil` before adding them to the hash is generally faster than using `compact!`

https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/693#discussion_r1360669417